src/cinder/block_store.go

```diff
@@ -0,0 +1,259 @@
+		return "", err
+	}
+	b.log.Infof("Snapshot is in 'available' state", snapshotName, snapshotID)
+
+	// Create Cinder Volume from snapshot (backup)
+	b.log.Infof("Starting to create volume from snapshot")
+	opts := volumes.CreateOpts{
+		Description:      "Velero backup from snapshot",
+		Name:             volumeName,
+		VolumeType:       volumeType,
+		AvailabilityZone: volumeAZ,
+		SnapshotID:       snapshotID,
+	}
+
+	var cinderVolume *volumes.Volume
+	cinderVolume, err = volumes.Create(b.client, opts).Extract()
+	if err != nil {
+		b.log.Errorf("failed to create volume from snapshot", snapshotName, snapshotID)
+		return "", errors.WithStack(err)
+	}
+	b.log.Infof("Backup volume was created", cinderVolume.ID)
+
+	// Remember the volume
+	b.volumes[cinderVolume.ID] = volumes.Volume{
+		Description:      "Velero backup from snapshot",
+		VolumeType:       volumeType,
+		AvailabilityZone: volumeAZ,
+		SnapshotID:       snapshotID,
+		ID:               cinderVolume.ID,
+		Name:             volumeName,
+	}
+
+	return cinderVolume.ID, nil
+}
+
+// GetVolumeInfo returns the type the specified volume in the given availability zone.
+// IOPS is not used as it is not supported by Cinder.
+func (b *BlockStore) GetVolumeInfo(volumeID, volumeAZ string) (string, *int64, error) {
+	b.log.Infof("GetVolumeInfo called", volumeID, volumeAZ)
+
+	// Volume exists in list of volumes
+	if vol, ok := b.volumes[volumeID]; ok {
+		return vol.VolumeType, nil, nil
+	}
+
+	// Volume doesn't exist in list of volumes
+	return "", nil, fmt.Errorf("volume %v not found", volumeID)
+}
+
+// IsVolumeReady Check if the volume is ready.
+func (b *BlockStore) IsVolumeReady(volumeID, volumeAZ string) (ready bool, err error) {
+	b.log.Infof("IsVolumeReady called", volumeID, volumeAZ)
+
+	// If volume doesn't exist in list of volumes
+	_, ok := b.volumes[volumeID]
+	if !ok {
+		return false, fmt.Errorf("volume %v doesn't exist in volumes map", volumeID)
+	}
+
+	// Get volume object from Cinder
+	cinderVolume, err := volumes.Get(b.client, volumeID).Extract()
+	if err != nil {
+		b.log.Errorf("failed to get volume %v from Cinder", volumeID)
+		return false, err
+	}
+
+	// These are ready states
+	//   https://github.com/openstack/cinder/blob/master/api-ref/source/v3/volumes-v3-volumes.inc#volumes-volumes
+	if cinderVolume.Status == "available" || cinderVolume.Status == "in-use" {
+		return true, nil
+	}
+
+	// Volume is not in one of the "ready" states
+	return false, fmt.Errorf("volume %v is not in ready state, the status is %v", volumeID, cinderVolume.Status)
+}
+
+// CreateSnapshot creates a snapshot of the specified volume, and applies any provided
+// set of tags to the snapshot.
+func (b *BlockStore) CreateSnapshot(volumeID, volumeAZ string, tags map[string]string) (string, error) {
+	b.log.Infof("CreateSnapshot called", volumeID, volumeAZ, tags)
+	snapshotName := volumeID + ".snap." + strconv.FormatUint(rand.Uint64(), 10)
```

Comment:
i'd rather use fmt.Sprintf

Reply:
Updated in ada815b5b82a4b608f745e2fe6df90abb22a6254